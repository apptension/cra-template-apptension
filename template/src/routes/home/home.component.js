import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';

import { FormattedMessage } from 'react-intl';
import messages from './home.messages';
import { LanguageSwitcher } from '../../shared/components/languageSwitcher';
import { Container, Logo } from './home.styles';
import { H1 } from '../../theme/typography';

export const HomeComponent = ({ formatMessage }) => (
  <Container>
    <Helmet title={formatMessage(messages.pageTitle)} />

    <H1>
      <FormattedMessage {...messages.welcome} />
    </H1>

    <Logo />

    <LanguageSwitcher />
  </Container>
);

HomeComponent.propTypes = {
  formatMessage: PropTypes.func.isRequired,
};
