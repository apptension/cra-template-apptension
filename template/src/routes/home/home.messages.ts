/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  pageTitle: {
    id: 'home.pageTitle',
    defaultMessage: 'Homepage',
  },
  welcome: {
    id: 'home.welcome',
    defaultMessage: 'Hello World!',
  },
});
