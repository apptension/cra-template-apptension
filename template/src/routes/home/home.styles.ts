import styled from 'styled-components';
import { ReactComponent as LogoSVG } from '../../images/icons/logo.svg';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
`;

export const Logo = styled(LogoSVG)`
  width: 100px;
  margin-bottom: 20px;
`;
