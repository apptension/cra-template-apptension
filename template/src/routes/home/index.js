import { hot } from 'react-hot-loader/root';
import { HomeComponent } from './home.component.tsx';

export const Home = hot(HomeComponent);
