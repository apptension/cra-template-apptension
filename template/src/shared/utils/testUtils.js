import React from 'react';
import PropTypes from 'prop-types';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Route, Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import { DEFAULT_LOCALE, translationMessages } from '../../i18n';
import { store as fixturesStore } from '../../../fixtures/store';

export * from '@testing-library/react';

export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = '/') => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const pushSpy = jest.spyOn(history, 'push');
  return {
    history,
    pushSpy,
  };
};

export const ProvidersWrapper = ({ children, context = {} }) => {
  const { router = {}, store = fixturesStore, messages } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  return (
    <Router history={routerHistory}>
      <HelmetProvider>
        <IntlProvider {...intlProviderMockProps}>
          <Provider store={createStore(identity, store)}>
            <Route path={routePath}>{children}</Route>
          </Provider>
        </IntlProvider>
      </HelmetProvider>
    </Router>
  );
};

ProvidersWrapper.propTypes = {
  context: PropTypes.shape({
    router: PropTypes.shape({
      url: PropTypes.string,
      routePath: PropTypes.string,
      history: PropTypes.object,
    }),
    store: PropTypes.object,
    messages: PropTypes.object,
  }),
};

export const makeContextRenderer = component => (props = {}, context = {}) =>
  render(component(props), {
    wrapper: ({ children }) => <ProvidersWrapper context={context}>{children}</ProvidersWrapper>,
  });

export const makePropsRenderer = component => (props = {}) => render(component(props));
