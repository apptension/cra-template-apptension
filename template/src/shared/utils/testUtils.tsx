import React, { ReactNode, ReactElement, FC } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { identity } from 'ramda';
import { Provider } from 'react-redux';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Route, Router } from 'react-router';
import { IntlProvider } from 'react-intl';
import produce from 'immer';

import { DEFAULT_LOCALE, translationMessages, MessagesObject } from '../../i18n';
import { store as fixturesStore } from '../../../fixtures/store';
import { ResponsiveThemeProvider } from '../components/responsiveThemeProvider';
import { theme as defaultTheme } from '../../theme/theme';
import { GlobalState } from '../../modules/reducers';
import { LOCALES_INITIAL_STATE } from '../../modules/locales';
import { STARTUP_INITIAL_STATE } from '../../modules/startup';
import { USERS_INITIAL_STATE } from '../../modules/users';
//<-- IMPORT MODULE STATE -->

const defaultGlobalState: GlobalState = {
  locales: LOCALES_INITIAL_STATE,
  startup: STARTUP_INITIAL_STATE,
  users: USERS_INITIAL_STATE,
  //<-- INJECT MODULE STATE -->
};

export const PLACEHOLDER_TEST_ID = 'content';
export const PLACEHOLDER_CONTENT = <span data-testid="content">content</span>;

export const spiedHistory = (route = '/') => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const pushSpy = jest.spyOn(history, 'push');
  return {
    history,
    pushSpy,
  };
};

interface ContextData {
  router?: {
    url?: string;
    routePath?: string;
    history?: MemoryHistory;
  };
  store?: GlobalState;
  messages?: MessagesObject;
  theme?: object;
}

interface ProvidersWrapperProps {
  children: ReactNode;
  context: ContextData;
}

export const ProvidersWrapper: FC<ProvidersWrapperProps> = ({ children, context = {} }) => {
  const { router = {}, store = fixturesStore, messages, theme = defaultTheme } = context;
  const { url = `/${DEFAULT_LOCALE}`, routePath = '/:lang/', history } = router;

  const routerHistory: MemoryHistory = history ?? createMemoryHistory({ initialEntries: [url] });

  const intlProviderMockProps = {
    locale: DEFAULT_LOCALE,
    messages: messages ?? translationMessages[DEFAULT_LOCALE],
  };

  return (
    <Router history={routerHistory}>
      <ResponsiveThemeProvider theme={theme}>
        <HelmetProvider>
          <IntlProvider {...intlProviderMockProps}>
            <Provider store={createStore(identity, store)}>
              <Route path={routePath}>{children}</Route>
            </Provider>
          </IntlProvider>
        </HelmetProvider>
      </ResponsiveThemeProvider>
    </Router>
  );
};

export const makeContextRenderer = <T, _>(component: (props: T | {}) => ReactElement) => (
  props?: T,
  context?: ContextData
) =>
  render(component(props ?? {}), {
    wrapper: ({ children }) => <ProvidersWrapper context={context}>{children}</ProvidersWrapper>,
  });

export const makePropsRenderer = <T, _>(component: (props: T | {}) => ReactElement) => (props?: T) =>
  render(component(props ?? {}));

export const prepareState = (stateSetter: (draftState: GlobalState) => void) =>
  produce(defaultGlobalState, stateSetter);
