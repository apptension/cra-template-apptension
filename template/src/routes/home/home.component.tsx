import React from 'react';
import { Helmet } from 'react-helmet-async';
import { FormattedMessage, useIntl } from 'react-intl';

import { LanguageSwitcher } from '../../shared/components/languageSwitcher';
import { H1 } from '../../theme/typography';
import { Users } from '../../shared/components/users';
import { Container, Logo } from './home.styles';

export const Home = () => {
  const intl = useIntl();

  return (
    <Container>
      <Helmet
        title={intl.formatMessage({
          defaultMessage: 'Homepage',
          description: 'Home / page title',
        })}
      />

      <H1>
        <FormattedMessage defaultMessage="Hello world!" description="Home / title" />
      </H1>

      <Logo />

      <LanguageSwitcher />

      <Users />
    </Container>
  );
};
