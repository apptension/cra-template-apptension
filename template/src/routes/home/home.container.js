import React from 'react';

import { useIntl } from 'react-intl';
import { HomeComponent } from './home.component';

export const HomeContainer = () => {
  const { formatMessage } = useIntl();

  const componentProps = {
    formatMessage,
  };

  return <HomeComponent {...componentProps} />;
};
