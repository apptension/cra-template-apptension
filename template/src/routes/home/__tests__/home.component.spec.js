import React from 'react';
import { Home } from '../home.component';
import { makeContextRenderer } from '../../../shared/utils/testUtils';

describe('Home: Component', () => {
  const defaultProps = {};

  const component = props => <Home {...defaultProps} {...props} />;
  const render = makeContextRenderer(component);

  it('should render correctly', () => {
    const { container } = render();
    expect(container.firstChild).toMatchSnapshot();
  });
});
