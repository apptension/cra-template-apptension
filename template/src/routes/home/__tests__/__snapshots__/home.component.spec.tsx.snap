// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home: Component should render correctly 1`] = `
.c1 {
  font-family: OpenSans,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 39px;
  line-height: 43px;
}

.c3 {
  padding: 8pxpx 16pxpx;
  border: 1px solid #0B0B0E;
  color: #42f272;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c2 {
  width: 100px;
  margin-bottom: 20px;
}

<div
  class="c0"
>
  <h1
    class="c1"
  >
    Hello world!
  </h1>
  <svg
    class="c2"
  >
    logo.svg
  </svg>
  <select
    class=""
  >
    <option
      value="en"
    >
      en
    </option>
    <option
      value="pl"
    >
      pl
    </option>
  </select>
  <div
    class=""
  >
    <button
      class="c3"
    >
      Fetch users
    </button>
    <ul />
  </div>
</div>
`;
